### DISCUSSION TITLE
[v5] Dynamically controlling the connection used in '@ioc:Adonis/Addons/Auth' (for multi-tenancy)

### DISCUSSION THREAD

**Discussion by vookio** [adonisjs/core#2314] [Help]
Hello,

Thanks for this great framework. After developing an app in v4.1, now developing new one in v5. It is going pretty well. 

My 2c: It would be great to have documentation sections dedicated to multi-tenancy and project structure for microservices.

My question: 
Having one global master db, and one tenant db for each customer.
What I want to achieve is dynamically controlling where to store auth token and where to do auth check as we do in lucid models by setting 'connection' or 'client'.

config/auth.ts file allows setting 'connection' in tokenProvider but I could not find a way to dynamically control it later.
 
```
...
tokenProvider: {
        driver: 'database',
        table: 'api_tokens',
        foreignKey: 'user_id',
        connection: 'tenant01',
        type: 'tenant',
      },
...
```
For example; how can I run the code below via connection 'tenant02' at any time, if it is defined as 'tenant01' in config/auth.ts as above? 
```
const token = await auth.use('master').attempt(email, password)

// or

const options = {...}
const token = await auth.use('tenant').generate(user, options)
```
Any code snippet would be lovely. In #1587, @incraigulous was mentioning similar thing I guess.

P.S. Setting 'connection' in options did not help. It expected the keys defined in "options" to be defined in "api_tokens" table.

**Comment by vookio** [comment]
I also checked if generating token without saving to db is possible so that I could manually insert into api_tokens table in the correct database, but I could not find a way. This would still require sending auth attempt to correct tenant db. 

`const token = await auth.use('tenant').generate(user, options)` this line tries inserting to db but wrong one.

**Reply by vookio** [reply]
After further checking, I think setConnection method can be exposed here for the usage below:

```
const token = await auth.use('tenant').connection(`tenant_${tenantId}`).attempt(email, password)

// or

const options = {...}
const token = await auth.use('tenant').connection(`tenant_${tenantId}`).generate(user, options)
```

@thetutlage @RomainLanz could you please help me on that? I could not figure out how to do it myself in past a few days. Thanks in advance.

**Reply by thetutlage** [reply]
Hello @vookio. So we do have some hooks in place and I think, we may need a few more to make it work with multi tenant apps.

Currently, we allow you to access the UserProvider using the `auth.use('tenant').provider` property and then you can use the hooks to modify the query.

```ts
auth.use('tenant').provider.before('findUser', () => {})
```

But I believe, we will have to expose the token provider as well to make it work. Currently, I am busy with the next release and I will get this added to that release (if you can want for a few weeks)

**Reply by vookio** [reply]
Thanks for your reply @thetutlage, I am looking forward to the release then. 🎉
In the meantime I will explore hooks.

**Reply by thetutlage** [reply]
Added here. Will be released soon. https://github.com/adonisjs/auth/commit/ea71af573c85373694d0ae75a5c88fa895df383b

You will be able to define custom connection as follows. 

```ts
const tenantGuard = auth.use('tenant')

tenantGuard.provider.setConnection('')
tenantGuard.tokenProvider.setConnection('')

// authenticate

await tenantGuard.attempt('uid', 'password')
```

**Reply by vookio** [reply]
Great news! I will update the project and try this as soon as docs are released.
